src/rdb_protocol/terms/string.cc

```diff
@@ -125,6 +127,55 @@ class match_term_t : public op_term_t {
     virtual const char *name() const { return "match"; }
 };
 
+class format_term_t : public op_term_t {
+public:
+    format_term_t(compile_env_t *env, const raw_term_t &term)
+            : op_term_t(env, term, argspec_t(2)) {}
+private:
+    virtual scoped_ptr_t <val_t> eval_impl(scope_env_t *env, args_t *args, eval_flags_t) const {
+        const std::string templ = args->arg(env, 0)->as_str().to_std();
+        const datum_t datum = args->arg(env, 1)->as_datum();
+
+        std::map <datum_string_t, datum_t> params{};
+        for (size_t i = 0; i < datum.obj_size(); ++i) {
+            const auto datum_pair{datum.get_pair(i)};
+            rcheck_target(&datum_pair.second, datum_pair.second.get_type() == datum_t::R_STR,
+                          base_exc_t::LOGIC,
+                          strprintf("Cannot call `%s` on objects of type `%s`.",
+                                    name(), datum_pair.second.get_type_name().c_str()));
+
+            params.insert(datum_pair);
+        }
+
+        std::string formatted{};
+        for (size_t pos = 0; pos < templ.size(); ++pos) {
+            if (templ[pos] == '{') {
+                size_t param_end {1};
+                std::string param {};
+
+                while (pos + param_end < templ.size() && templ[pos + param_end] != '}') {
+                    param.push_back(templ[pos + param_end]);
+                    ++param_end;
+                }
+
+                const datum_string_t datum_param {datum_string_t(param)};
+                if (params.count(datum_param)) {
+                    formatted += params.at(datum_param).as_str().to_std();
+                } else {
+                    formatted += param;
```

Comment:
This is highly questionable fallback behavior.  It think we should throw an error if the field is absent.

If some kind of field-or-fallback-if-doesn't-exist logic is desired, it could be specified in the format syntax, or the user could r.merge the document into an object with default values.

Reply:
That's a fair point. I'll refactor it to throw an error instead.